### ISSUE TITLE
There is no way to choose IPv4 preferably to IPv6

### ISSUE THREAD

**Issue by ghost** [elad/gnats-nb#630]
### Metadata
- Number: [22761](http://gnats.netbsd.org/22761)
- Originator: Todd Vierling &lt;[tv@duh.org](mailto:tv@duh.org)&gt;
- Arrived: Fri Sep 12 17:34:00 +0000 2003
- Last modified: Sat Sep 13 22:23:00 +0000 2003
- Environment: System: NetBSD server.duh.org 1.6.1_STABLE NetBSD 1.6.1_STABLE (SERVER) #1: Fri Sep 12 11:28:14 EDT 2003 tv@server.duh.org:/export/SRC/duh/netbsd-kernels/SERVER i386 Architecture: i386 Machine: i386
### Description

```


Many times I've been bitten by this, and I'm tired of typing "ftp -4"
all the time just because I'm trying to contribute to the deployment
of IPv6 but don't have the tunnel bandwidth to do major-throughput
operations over v6 on a daily basis.

My configuration is a 6to4 tunnel, which has pretty high RTT thanks to the
fact that very few providers advertise an exit point to 2002::/16 from the
6bone.  (Unfortunately, this situation is something I cannot fix.)


```
### How to repeat

```


Hook up a 6to4 tunnel, or any other relatively high latency IPv6-over-IPv4
tunnel.  Go to any site that has both IPv6 and IPv4 addresses.  See that
the IPv6 address family and address is always picked first, and that there
is no way to get around this without telling the network tool to use IPv4
explicitly.


```
### Fix

```


NFC.  I'd presume that the first place this should be configurable is in
getaddrinfo(3).  It always returns IPv6 first, so perhaps it can be made
to return IPv4 first.  This might solve this whole issue altogether for
most tools that simply depend on what getaddrinfo(3) returns.

Other programs that do explicit IPv6 and IPv4 lookups separately would
need access to the same config option, likely via a function call, and
make use of it.

```
### Audit trail

```


From: fredb@immanent.net (Frederick Bruckman)
To: tv@duh.org
Cc: fredb@netbsd.org, itojun@netbsd.org, gnats-bugs@gnats.netbsd.org
Subject: Re: bin/22761: There is no way to choose IPv4 preferably to IPv6
Date: Sat, 13 Sep 2003 08:52:14 -0500 (CDT)

 In article <200309121732.h8CHWWtL016506@server.duh.org>,
    tv@duh.org writes:
 > 
 >>Synopsis:       There is no way to choose IPv4 preferably to IPv6

 > NFC.  I'd presume that the first place this should be configurable is in
 > getaddrinfo(3).  It always returns IPv6 first, so perhaps it can be made
 > to return IPv4 first.  This might solve this whole issue altogether for
 > most tools that simply depend on what getaddrinfo(3) returns.

 RFC 3484 has something to say about this. To be compliant, we'll
 ultimately need some host-wide configuration file for this. When
 I last asked about it, itojun said that KAME has it, but there
 were problems merging the code into NetBSD:

     http://mail-index.netbsd.org/tech-net/2003/07/14/0000.html
     http://mail-index.netbsd.org/tech-net/2003/07/13/0002.html

 Frederick

From: itojun@itojun.org (Jun-ichiro itojun Hagino)
To: fredb@immanent.net
Cc: tv@duh.org, fredb@netbsd.org, gnats-bugs@gnats.netbsd.org
Subject: Re: bin/22761: There is no way to choose IPv4 preferably to IPv6
Date: Sun, 14 Sep 2003 07:22:56 +0900 (JST)

 > In article <200309121732.h8CHWWtL016506@server.duh.org>,
 >  tv@duh.org writes:
 > > 
 > >>Synopsis:       There is no way to choose IPv4 preferably to IPv6
 >  
 > > NFC.  I'd presume that the first place this should be configurable is in
 > > getaddrinfo(3).  It always returns IPv6 first, so perhaps it can be made
 > > to return IPv4 first.  This might solve this whole issue altogether for
 > > most tools that simply depend on what getaddrinfo(3) returns.
 > 
 > RFC 3484 has something to say about this. To be compliant, we'll
 > ultimately need some host-wide configuration file for this. When
 > I last asked about it, itojun said that KAME has it, but there
 > were problems merging the code into NetBSD:
 > 
 >     http://mail-index.netbsd.org/tech-net/2003/07/14/0000.html
 >     http://mail-index.netbsd.org/tech-net/2003/07/13/0002.html

    basically it needs some cleanup.  i'll try to work on it sooner
    (now my work-stack is full :-).

 itojun

```
